File: src/main/java/sh/ball/audio/effect/VectorCancellingEffect.java
package sh.ball.audio.effect;

import sh.ball.shapes.Vector2;

public class VectorCancellingEffect implements SettableEffect {

  private double frequency = 1;
  private double nextInvert;
  private int lastCount = 0;

  public VectorCancellingEffect() {
  }

  @Override
  public void setValue(double value) {
    this.frequency = 1.0 + 9.0 * value;
  }

  @Override
  public Vector2 apply(int count, Vector2 v) {
    if (count < lastCount) {
      nextInvert = nextInvert - lastCount + frequency;
    }
    lastCount = count;
    if (count >= nextInvert) {
      nextInvert += frequency;
      return v;
    } else {
      return v.scale(-1);
    }
  }
}

File: Source/audio/VectorCancellingEffect.cpp
#include "VectorCancellingEffect.h"

VectorCancellingEffect::VectorCancellingEffect() {}

VectorCancellingEffect::~VectorCancellingEffect() {}

osci::Point VectorCancellingEffect::apply(int index, osci::Point input, const std::vector<std::atomic<double>>& values, double sampleRate) {
    double value = values[0];
    if (value < 0.001) {
		return input;
    }
    double frequency = 1.0 + 9.0 * value;
    if (index < lastIndex) {
        nextInvert = nextInvert - lastIndex + frequency;
    }
    lastIndex = index;
    if (index >= nextInvert) {
        nextInvert += frequency;
    } else {
        input.scale(-1, -1, 1);
    }
    return input;
}
